src/isar/models/mission.py

```diff
@@ -1,43 +1,66 @@
-from dataclasses import dataclass
+from dataclasses import dataclass, field
 from datetime import datetime
-from typing import List, Union
-from uuid import UUID
+from typing import Iterator, List, Union
+from uuid import UUID, uuid4
 
 from isar.config.settings import settings
 from isar.models.mission_metadata.mission_metadata import MissionMetadata
-from robot_interface.models.mission import STEPS
-from robot_interface.models.mission.step import DriveToPose, InspectionStep
+from robot_interface.models.mission import STEPS, Step, StepStatus
 
 
 @dataclass
-class Mission:
+class Task:
     steps: List[STEPS]
+    status: StepStatus = field(default=StepStatus.NotStarted, init=False)
+    id: UUID = field(default_factory=uuid4, init=False)
+    iterator: Iterator = None
+
+    def next_step(self) -> Step:
+        return self.iterator.__next__()
+
+    def is_finished(self) -> bool:
+        for step in self.steps:
+            if step.status is StepStatus.Failed or step.status is StepStatus.Unexpected:
+                # One step has failed meaning the task as a whole should be aborted
+                self.status = StepStatus.Failed
```

Comment:
There is a difference between types of steps here. If a DriveToPose step fails, I agree that the task as a whole should fail. But if a task consisting of [DriveToPose, TakeImage and TakeThermalImage] and the TakeThermalImage fails, the TakeImage could still succeed and the task could be partly successful

Reply:
That is a good point. What is the best way to deal with it though?

Should a task conceptually be defined as a "requires all to pass for task to pass"? Because that is much easier to define and work with in practice. This would require that we split the `TakeImage` and `TakeThermalImage` in two tasks in your example to circumvent the issue though.

Reply:
Yes, having a `required` attribute on the required steps could be a good solution.

Not sure if this is possible to set on the type of task, i.e. is a DriveToPose always required, and is a TakeImage always optional?

Reply:
Then again, if you had only a `DriveToPose` and `TakeImage` you would say the `Task` failed if the `TakeImage` step which isn't listed as required failed.

I would say a `Task` has failed even if one out of X inspections have failed.

We could implement something along the line of

```
if MotionStep has failed:
    task has failed
elif InspectionStep has failed and steps remaining:
    task has partly failed but still perform remaining steps
```

And some more logic probably.

Reply:
Based on the discussion, I see three options: 

1. Leave the implementation as it is, requiring all steps to succeed for a  `Task` to succeed. The drawback with this solution is that if the task have several inspection steps, some of them might have been able to succeed.

2. Implement one `Task` per `InspectionStep`.

3. Let the `Task` fail if a `MotionStep` fails. This allows all `InspectionsStep`s to be executed. However, what will then be the requirement for the `Task` to succeed? At least one successful `InspectionStep`?

What are your thoughts, @aestene @tsundvoll?

Reply:
This is related to dependencies between steps, right?
If this is the case, I think a task should only fail if it does not make sense to continue the task. (Closest to option 3 from  @anetteu ). Don't abort the task if a measure step fails, but abort the task if a drivetopose step fails

Reply:
A PartialSuccessful status could be used to indicate partially successful tasks where one or more measure steps have failed.

Reply:
There is also a difference between aborting the task and considering the task as failed. I would say nr. 3 with information to the user that the task has only partly succeeded. Should we add a new possible status for a task? `Status.PartlySuccessful`

Reply:
This has now been implemented in such a way that a task will be `PartiallySuccessful` if one or more `InspectionSteps` have failed. If all `InspectionSteps` have failed the entire task will be considered as failed.

If a `MotionStep` fails the entire task is considered as failed. There is an underlying assumption here that most tasks will only contain one `MotionStep` which I think is fair.

Reply:
When/if we wish to deal with multiple inspections of the same target from different angles we'll look into this some more.